File: shared/database/src/test-connection.ts
#!/usr/bin/env node
/**
 * RuvVector Database Connection Test
 * Tests connectivity from all integrations to the RuvVector Postgres instance
 */

import { RuvectorDatabase, getDefaultConfig, ConnectionTestResult } from './index.js';

interface IntegrationTestResult {
  integration: string;
  result: ConnectionTestResult;
}

const INTEGRATIONS = [
  'anthropic',
  'aws/s3',
  'aws/ses',
  'cohere',
  'gemini',
  'github',
  'google-drive',
  'groq',
  'mistral',
  'oauth2',
  'openai',
  'slack',
  'smtp',
];

async function testIntegrationConnectivity(
  integrationName: string,
  db: RuvectorDatabase
): Promise<IntegrationTestResult> {
  console.log(`\nTesting: ${integrationName}`);
  console.log('-'.repeat(40));

  const result = await db.testConnection();

  if (result.success) {
    console.log(`  Status: PASS`);
    console.log(`  Database: ${result.details?.database}`);
    console.log(`  User: ${result.details?.user}`);
    console.log(`  Extensions: ${result.details?.extensions?.join(', ') || 'none'}`);
  } else {
    console.log(`  Status: FAIL`);
    console.log(`  Error: ${result.message}`);
  }

  return { integration: integrationName, result };
}

async function main(): Promise<void> {
  console.log('='.repeat(60));
  console.log('RuvVector Database Connectivity Test');
  console.log('='.repeat(60));

  const config = getDefaultConfig();
  console.log(`\nConnection Target:`);
  console.log(`  Host: ${config.host}`);
  console.log(`  Port: ${config.port}`);
  console.log(`  Database: ${config.database}`);
  console.log(`  User: ${config.user}`);

  const db = new RuvectorDatabase(config);
  const results: IntegrationTestResult[] = [];

  try {
    for (const integration of INTEGRATIONS) {
      const testResult = await testIntegrationConnectivity(integration, db);
      results.push(testResult);
    }

    // Summary
    console.log('\n' + '='.repeat(60));
    console.log('SUMMARY');
    console.log('='.repeat(60));

    const passed = results.filter((r) => r.result.success);
    const failed = results.filter((r) => !r.result.success);

    console.log(`\nTotal: ${results.length}`);
    console.log(`Passed: ${passed.length}`);
    console.log(`Failed: ${failed.length}`);

    if (failed.length > 0) {
      console.log('\nFailed integrations:');
      for (const f of failed) {
        console.log(`  - ${f.integration}: ${f.result.message}`);
      }
      process.exit(1);
    }

    console.log('\nAll integrations can connect to RuvVector Postgres!');
  } finally {
    await db.close();
  }
}

main().catch((err) => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: shared/database-rs/src/bin/test_connection.rs
//! RuvVector Database Connection Test
//! Tests connectivity from all Rust integrations to the RuvVector Postgres instance

use integrations_database::{DatabaseConfig, RuvectorDatabase};

const INTEGRATIONS: &[&str] = &[
    "anthropic",
    "aws/s3",
    "aws/ses",
    "cohere",
    "gemini",
    "github",
    "google-drive",
    "groq",
    "mistral",
    "oauth2",
    "openai",
    "slack",
    "smtp",
];

#[tokio::main]
async fn main() {
    println!("{}", "=".repeat(60));
    println!("RuvVector Database Connectivity Test (Rust)");
    println!("{}", "=".repeat(60));

    let config = DatabaseConfig::default();
    println!("\nConnection Target:");
    println!("  Host: {}", config.host);
    println!("  Port: {}", config.port);
    println!("  Database: {}", config.database);
    println!("  User: {}", config.user);

    let db = match RuvectorDatabase::new(config).await {
        Ok(db) => db,
        Err(e) => {
            eprintln!("\nFailed to create database pool: {}", e);
            std::process::exit(1);
        }
    };

    let mut passed = 0;
    let mut failed = 0;
    let mut failures: Vec<(String, String)> = Vec::new();

    for integration in INTEGRATIONS {
        println!("\nTesting: {}", integration);
        println!("{}", "-".repeat(40));

        let result = db.test_connection().await;

        if result.success {
            println!("  Status: PASS");
            println!(
                "  Database: {}",
                result.database.as_deref().unwrap_or("unknown")
            );
            println!("  User: {}", result.user.as_deref().unwrap_or("unknown"));
            println!("  Extensions: {}", result.extensions.join(", "));
            passed += 1;
        } else {
            println!("  Status: FAIL");
            println!("  Error: {}", result.message);
            failures.push((integration.to_string(), result.message));
            failed += 1;
        }
    }

    println!("\n{}", "=".repeat(60));
    println!("SUMMARY");
    println!("{}", "=".repeat(60));
    println!("\nTotal: {}", INTEGRATIONS.len());
    println!("Passed: {}", passed);
    println!("Failed: {}", failed);

    if !failures.is_empty() {
        println!("\nFailed integrations:");
        for (name, msg) in &failures {
            println!("  - {}: {}", name, msg);
        }
        std::process::exit(1);
    }

    println!("\nAll integrations can connect to RuvVector Postgres!");
}
